src/test/app/Batch_test.cpp

```diff
@@ -0,0 +1,3047 @@
+            BEAST_EXPECT(env.balance(alice) == preAlice - batchFee);
+            BEAST_EXPECT(env.balance(bob) == preBob);
+        }
+    }
+
+    void
+    testOnlyOne(FeatureBitset features)
+    {
+        testcase("only one");
+
+        using namespace test::jtx;
+        using namespace std::literals;
+
+        test::jtx::Env env{*this, envconfig()};
+
+        auto const alice = Account("alice");
+        auto const bob = Account("bob");
+        auto const gw = Account("gw");
+        auto const USD = gw["USD"];
+        env.fund(XRP(1000), alice, bob, gw);
+        env.close();
+
+        // tec failure
+        {
+            auto const preAlice = env.balance(alice);
+            auto const preBob = env.balance(bob);
+
+            auto const batchFee = batch::calcBatchFee(env, 0, 3);
+            auto const seq = env.seq(alice);
+            env(batch::outer(alice, seq, batchFee, tfOnlyOne),
+                batch::inner(pay(alice, bob, XRP(999)), seq + 1),
+                batch::inner(pay(alice, bob, XRP(1)), seq + 2),
+                batch::inner(pay(alice, bob, XRP(2)), seq + 3),
+                ter(tesSUCCESS));
+            auto const txIDs = env.tx()->getBatchTransactionIDs();
+            TxID const parentBatchId = env.tx()->getTransactionID();
+            std::vector<TestBatchData> testCases = {
+                {"tecUNFUNDED_PAYMENT", to_string(txIDs[0])},
+                {"tesSUCCESS", to_string(txIDs[1])},
+            };
+            env.close();
+            validateBatch(env, parentBatchId, testCases);
+
+            // Alice consumes sequences (# of txns)
+            BEAST_EXPECT(env.seq(alice) == seq + 3);
+
+            // Alice pays XRP & Fee; Bob receives XRP
+            BEAST_EXPECT(env.balance(alice) == preAlice - XRP(1) - batchFee);
+            BEAST_EXPECT(env.balance(bob) == preBob + XRP(1));
+        }
+
+        // non tec failure
+        {
+            auto const preAlice = env.balance(alice);
+            auto const preBob = env.balance(bob);
+
+            auto const batchFee = batch::calcBatchFee(env, 0, 3);
+            auto const seq = env.seq(alice);
+            auto tx1 = trust(alice, USD(1000), tfSetfAuth);
+            tx1[jss::Fee] = "0";
+            tx1[jss::Sequence] = seq + 1;
+            tx1[jss::SigningPubKey] = "";
+            tx1[jss::Flags] = tx1[jss::Flags].asUInt() | tfInnerBatchTxn;
+            env(batch::outer(alice, seq, batchFee, tfOnlyOne),
+                batch::inner_nofill(tx1),
+                // duplicate txn id
+                batch::inner(pay(alice, bob, XRP(1)), seq + 1),
+                batch::inner(pay(alice, bob, XRP(2)), seq + 3),
+                ter(tesSUCCESS));
+            auto const txIDs = env.tx()->getBatchTransactionIDs();
+            TxID const parentBatchId = env.tx()->getTransactionID();
+            std::vector<TestBatchData> testCases = {
+                // tx #1 fails with tefNO_AUTH_REQUIRED
```

Comment:
There should be a check that this tx isn't included in the ledger.

Reply:
I think the best way to solve this would be to modify the framework to handle failures as well.